File: 31-next-permutation/next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int Break_Point = -1;
        int n= nums.size();
        for(int i=n-2;i>=0;i--){
            if(nums[i]<nums[i+1]){
                Break_Point=i;
                break;
            }
        }
        for(int i=n-1; i>=Break_Point && Break_Point != -1; i--){
            if(nums[i] > nums[Break_Point]){
                swap(nums[i], nums[Break_Point]);
                break;
            }
        }
        reverse(nums.begin() + Break_Point + 1, nums.end());
    }
};

File: 31-next-permutation/next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        // step-1 find the pivot
        int pivot = -1;
        for (int i = nums.length - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                pivot = i;
                break;
            }
        }
        // step-2 find right most element greater than pivot
        if (pivot != -1) {
            for (int i = nums.length - 1; i >= 0; i--) {
                if (nums[i] > nums[pivot]) {
                    int temp = nums[i];
                    nums[i] = nums[pivot];
                    nums[pivot] = temp;
                    break;
                }
            }
        }
        // reverse the remaining part
        int start = pivot + 1;
        int n = nums.length;
        while (start <= n - 1) {
            int temp = nums[start];
            nums[start] = nums[n - 1];
            nums[n - 1] = temp;
            start++;
            n--;
        }
    }
}